src/server/session.ts

```diff
@@ -1199,10 +1205,20 @@ namespace ts.server {
 
         private getCompletionEntryDetails(args: protocol.CompletionDetailsRequestArgs): ReadonlyArray<protocol.CompletionEntryDetails> {
             const { file, project } = this.getFileAndProject(args);
-            const position = this.getPositionInFile(args, file);
+            const scriptInfo = this.projectService.getScriptInfoForNormalizedPath(file);
+            const position = this.getPosition(args, scriptInfo);
+            const formattingOptions = project.projectService.getFormatCodeOptions(file);
 
-            return mapDefined(args.entryNames, entryName =>
-                project.getLanguageService().getCompletionEntryDetails(file, position, entryName));
+            return mapDefined(args.entryNames, entryName => {
+                const details = project.getLanguageService().getCompletionEntryDetails(file, position, entryName, formattingOptions);
+                if (details) {
+                    const mappedCodeActions = map(details.codeActions, action => this.mapCodeAction(action, scriptInfo));

```

Comment:
i would check if details.codeActions is not defined, and return details.

Reply:
or conisder:

```ts
if (details) {
    details.codeActions = map(details.codeActions, action => this.mapCodeAction(action, scriptInfo));
}
```

this avoids the extra object allocation, and will not be marshaled to the caller, since JSON will not marshal undefined properties.

Reply:
There's no type-safe way to mutate an object of one type into an object of a different type -- I would have to use `any` and that would cause errors if we needed to transform some other part of `details` and forgot about it. I don't think the extra allocation is a big deal, since it's only a single object per `completionEntryDetails` call.

Reply:
The ```getCodeFix``` and ```GetEditsForRefactor``` have option of  getting simplified result through those methods and full result through ```GetCodeFixesFull``` and ```GetEditsForRefactorFull```. Would we want something similar here too?

Reply:
Also I agree with @mhegazy we shouldnt be allocating object if there is no need. Also looking at this method it looks like there could be multiple entries requested which means those many more new objects?

Reply:
The simplified results appear to be trying to reduce the amount of JSON sent? That would be more important for "completions" (which has a simplified version) than for "completionDetails".
The `mapDefined` is in case the client makes multiple requests together, as in trying to get details on "foo" and "fop" at the same time. I don't think it would ever be in the thousands. I think there's a genuine need for a new object since it would have a different type -- see `decorateNavigationTree` for a similar function that returns an object of different type from its input.